Allow {symbol} interpolation outside of strings

Rangi42 opened the pull request:
Fixes #629

Closes #631

Rangi42 commented:
Now `peek` is solely responsible for doing symbol interpolation outside of strings. (`readString` has its own `readInterpolation` call because its `peek` calls need to have interpolation disabled.)

ISSOtm commented:
Why do strings need interpolation disabled?

ISSOtm commented:
Alright, I'm not a fan of this special case, but it's necessary for compatibility, and the implementation is acceptable.

[Checkpatch notes two style mismatches](https://github.com/gbdev/rgbds/pull/634/checks#step:4:13).

Finally, as this is a touchy part of the codebase, I'll page @AntonioND for a second pair of eyes; if he's satisfied with the change as well, he can merge this.

JL2210 commented on src/asm/lexer.c:
```diff
@@ -767,16 +769,23 @@ static int peekInternal(uint8_t distance)
-	int c = peekInternal(distance);
+	int c;
+
+restart:
```
I would use recursion or a loop for this instead.

Rangi42 replied:
@ISSOtm said he preferred `goto restart` to `continue` in a `for (;;)` loop. If tail call optimization can be relied upon, recursive `return peek(distance)` may be neater.

ISSOtm replied:
I would prefer to avoid relying on tail calling for this.

JL2210 replied:
It's reasonable to expect a compiler to perform tail-call elimination on something this simple.

ISSOtm replied:
We may see to that when we come to a round of performance tuning.

Rangi42 commented:
This, plus more printf-like format specifiers, will really simplify pokered's code.

```
TechnicalMachines:
n = 1
REPT NUM_TMS
	db TM{02d:n}_MOVE
n = n + 1
ENDR
```

```
;\1 source map
def_warps_to: MACRO
; output and purge each _WARP_TO_NUM_<N> warp_to, from N=1 to _NUM_WARPS
_WARP_TO_WIDTH = \1_WIDTH ; used in each _WARP_TO_NUM_<N>
n = 1
	REPT _NUM_WARPS
		_WARP_TO_NUM_{d:n}
		PURGE _WARP_TO_NUM_{d:n}
n = n + 1
	ENDR
ENDM
```

The "output and purge" comment there is basically self-explanatory.

AntonioND commented:
Ooops, I didn't see the review request. I've received so many notifications from this repo in the last few days that I've just started to get rid of most of them...

My problem with this PR is that it is just not clear what it does. There is no documentation and the only test is so long and complicated that is not useful as an example.

ISSOtm commented:
The lack of documentation is certainly a problem. I would define the content of this PR as "expanding the contents of symbols, like in strings, but in code".

@AntonioND If you're receiving too many notifications, you may want to tune your "Watch" settings?

AntonioND commented:
> The lack of documentation is certainly a problem. I would define the content of this PR as "expanding the contents of symbols, like in strings, but in code".

Yeah, true.

> @AntonioND If you're receiving too many notifications, you may want to tune your "Watch" settings?

Well, most of the time this isn't a problem because there isn't that much activity, and I like the instant notifications in that case. In situations like the current one, when there are so many changes, just ping me on Discord to make sure I see it.

Rangi42 commented:
I've simplified the test case file and added it to the `man` documentation. It covers the possible uses of `{interpolation}`:

- Applied to strings
- Applied to numbers
- Nested
- Using format specifiers
- Interpolating within ordinary code context
- Interpolating within a numeric literal
- Interpolating within a name to be defined
- Interpolating within a name to be purged

Please let me know if the example should be edited further.

AntonioND reviewed (approved):
Yeah, that's better.

ISSOtm commented:
Alright, I'm OK with the new changes as well, so we'll (finally!) merge this.
